cranelift/frontend/src/frontend.rs

```diff
@@ -791,28 +804,65 @@ impl<'a> FunctionBuilder<'a> {
         //    simplest thing. Besides, none of our mid-end optimization passes
         //    have run at this point in time yet, so there probably isn't much,
         //    if any, dead code.
-        for block in self.func_ctx.dfs.post_order_iter(&self.func) {
+        for block in self
+            .func_ctx
+            .dfs
+            .post_order_iter(&self.func)
+            // We have to `collect` here to release the borrow on `self.func` so
+            // we can add the stack map entries below.
+            .collect::<Vec<_>>()
+        {
+            log::trace!("stack map liveness: traversing {block:?}");
             let mut option_inst = self.func.layout.last_inst(block);
             while let Some(inst) = option_inst {
                 // (1) Remove values defined by this instruction from the `live`
                 // set.
                 for val in self.func.dfg.inst_results(inst) {
-                    live.remove(val);
+                    process_def(
+                        &self.func,
+                        &stack_slots,
+                        &mut free_stack_slots,
+                        &mut live,
+                        *val,
+                    );
                 }
 
-                // (2) If this instruction is a call, then we need to add a
-                // safepoint to record any values in `live`.
+                // (2) If this instruction is a safepoint, then we need to add
+                // stack map entries to record the values in `live`.
                 let opcode = self.func.dfg.insts[inst].opcode();
                 if opcode.is_call() && !opcode.is_return() {
-                    let mut live: SmallVec<[_; 4]> = live.iter().copied().collect();
-                    for chunk in live_vals_by_size(&self.func.dfg, &mut live) {
-                        let index = log2_size(&self.func.dfg, chunk[0]);
-                        max_vals_in_stack_map_by_log2_size[index] =
-                            core::cmp::max(max_vals_in_stack_map_by_log2_size[index], chunk.len());
+                    log::trace!(
+                        "stack map liveness: found safepoint: {inst:?}: {}",
+                        self.func.dfg.display_inst(inst)
+                    );
+                    log::trace!("  live = {live:?}");
+
+                    for val in &live {
+                        let ty = self.func.dfg.value_type(*val);
+                        let slot = *stack_slots.entry(*val).or_insert_with(|| {
+                            let size = self.func.dfg.value_type(*val).bytes();
```

Comment:
perhaps abstract out to an analogous `process_use` lambda here to make things a bit clearer, as above for defs?